### PR TITLE
Update network profiler to support dart:io HTTP profiling service extensions

### PR DESCRIPTION
**Pull request by bkonyi**

### CODE REVIEW

File: packages/devtools_app/lib/src/http/http_request_data.dart [hunk 1]
```diff
@@ -361,3 +448,164 @@ class HttpRequestData extends NetworkRequest {
...
+    int timelineMicrosBase,
+    this._request,
+  ) : super(timelineMicrosBase) {
+    if (_request.isResponseComplete) {
+      serviceManager.service
+          .getHttpProfileRequest(
+            _request.isolateId,
+            _request.id,
+          )
+          .then((value) => _request = value);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/http/http_request_data.dart:463]: call getFullRequestData here instead of duplicating the code

File: packages/devtools_app/lib/src/http/http_request_data.dart [hunk 2]
```diff
@@ -361,3 +448,164 @@ class HttpRequestData extends NetworkRequest {
...
+      if (!_request.isResponseComplete) return null;
+      if (_responseBody != null) return _responseBody;
+      _responseBody = utf8.decode(fullRequest.responseBody);
+      if (contentType.contains('json')) {
+        print('formatting json');
+        _responseBody = const JsonEncoder.withIndent('  ').convert(json.decode(_responseBody));
+      }
+      return _responseBody;
+    } on FormatException {
+      return '<binary data>'; //fullRequest.responseBody.toString();
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/http/http_request_data.dart:595]: remove comment?

File: packages/devtools_app/lib/src/http/http_request_data.dart [hunk 3]
```diff
@@ -361,3 +448,164 @@ class HttpRequestData extends NetworkRequest {
...
+    if (_request is! HttpProfileRequest) {
+      return null;
+    }
+    final fullRequest = _request as HttpProfileRequest;
+    try {
+      if (!_request.isResponseComplete) return null;
+      if (_responseBody != null) return _responseBody;
+      _responseBody = utf8.decode(fullRequest.responseBody);
+      if (contentType.contains('json')) {
+        print('formatting json');
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/http/http_request_data.dart:590]: remove print?

File: packages/devtools_app/lib/src/http/http_request_data.dart [hunk 4]
```diff
@@ -361,3 +448,164 @@ class HttpRequestData extends NetworkRequest {
...
+        _responseBody = const JsonEncoder.withIndent('  ').convert(json.decode(_responseBody));
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/http/http_request_data.dart:591]: Let's use `FormattedJson.encoder` to make sure we use the same encoder throughout DevTools. Can you also add a TODO that says consider returning a `Map<String, dynamic>` from responseBody so the `FormattedJson` widget can be used directly in the UI?

File: packages/devtools_app/lib/src/network/network_controller.dart [hunk 5]
```diff
@@ -171,6 +209,22 @@ class NetworkController
       currentValues.add(webSocket);
     }
 
+    if (timeline != null) {
+      _processTimeline(
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/network/network_controller.dart:213]: in what case will `timeline != null`? is this for backwards compatibility?

File: packages/devtools_app/lib/src/network/network_request_inspector_views.dart [hunk 6]
```diff
@@ -450,6 +451,25 @@ class NetworkRequestOverviewView extends StatelessWidget {
...
+        child: Tooltip(
+          waitDuration: tooltipWait,
+          message: 'Response - ${msText(duration)}',
+          child: Container(
+            height: _timingGraphHeight,
+            color: _nextColor(),
+          ),
+        ),
+      ),
+    );
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/network/network_request_inspector_views.dart:472]: this is duplicated above. put into a helper that takes a duration and a label.

File: packages/devtools_app/lib/src/network/network_service.dart [hunk 7]
```diff
@@ -2,6 +2,7 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:devtools_app/src/version.dart';
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/network/network_service.dart:5]: nit: use relative imports

File: packages/devtools_app/lib/src/vm_service_wrapper.dart [hunk 8]
```diff
@@ -503,6 +503,40 @@ class VmServiceWrapper implements VmService {
...
   }
 
+  /// The `getHttpProfile` RPC is used to retrieve HTTP profiling information
+  /// for requests made via `dart:io`'s `HttpClient`.
+  ///
+  /// The returned [HttpProfile] will only include requests issued after
+  /// [httpTimelineLogging] has been enabled or after the last
+  /// [clearHttpProfile] invocation.
+  Future<HttpProfile> getHttpProfile(String isolateId, {int updatedSince}) {
+    return trackFuture(
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/vm_service_wrapper.dart:513]: don't we need to add `isProtocolVersionSupported` checks around all of these to make sure we don't try to call this method on an old version of the vm_service?

File: packages/devtools_app/lib/src/network/network_service.dart [hunk 9]
```diff
@@ -39,18 +40,43 @@ class NetworkService {
...
+    assert(serviceManager.service != null);
+    if (serviceManager.service == null) return [];
+
+    final requests = <HttpProfileRequest>[];
+    await serviceManager.service.forEachIsolate((isolate) async {
+      final request = await serviceManager.service.getHttpProfile(
+        isolate.id,
+        updatedSince: networkController.lastRefreshMicros,
+      );
+      requests.addAll(request.requests);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/network/network_service.dart:75]: can we just return request.requests instead of making a copy of the list?

File: packages/devtools_app/lib/src/http/http_request_data.dart (hunk 3 above)
Comment (bkonyi) [reply packages/devtools_app/lib/src/http/http_request_data.dart:590]: Will do. I meant to remove this change for this PR and land it separately. WDYT?

File: packages/devtools_app/lib/src/network/network_controller.dart (hunk 5 above)
Comment (bkonyi) [reply packages/devtools_app/lib/src/network/network_controller.dart:213]: Yes, this is for backward compatibility. When we start shipping a pinned DevTools with the SDK we can tear this all out.

File: packages/devtools_app/lib/src/network/network_service.dart (hunk 7 above)
Comment (bkonyi) [reply packages/devtools_app/lib/src/network/network_service.dart:5]: I'll blame VSCode for this... :-)

File: packages/devtools_app/lib/src/network/network_service.dart (hunk 9 above)
Comment (bkonyi) [reply packages/devtools_app/lib/src/network/network_service.dart:75]: No, since we need to get profiling data from each isolate, which each returns their own `request`.

File: packages/devtools_app/lib/src/vm_service_wrapper.dart (hunk 8 above)
Comment (bkonyi) [reply packages/devtools_app/lib/src/vm_service_wrapper.dart:513]: Good call, yes we do.

File: packages/devtools_app/lib/src/network/network_service.dart (hunk 9 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/network/network_service.dart:75]: oops - didn't see `forEachIsolate`. sg.

File: packages/devtools_app/lib/src/http/http_request_data.dart (hunk 3 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/http/http_request_data.dart:590]: remove the json formatting change? Sure, you can land as a follow up if you want.

File: packages/devtools_app/lib/src/http/http_request_data.dart (hunk 4 above)
Comment (bkonyi) [reply packages/devtools_app/lib/src/http/http_request_data.dart:591]: Not sure we'd be able to return `Map<String, dynamic>` in the general case since a simple string response is completely valid.

File: packages/devtools_app/lib/src/network/network_request_inspector_views.dart [hunk 10]
```diff
@@ -420,6 +420,7 @@ class NetworkRequestOverviewView extends StatelessWidget {
     final _colors = [
       searchMatchColor,
       mainRasterColor,
+      mainAsyncColor,
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/network/network_request_inspector_views.dart:423]: This is something else I noticed. When we have only two colors available here and we encounter one really short event, there's no visible difference between the events on either side of the effectively invisible event. Having more colors here should help alleviate that particular issue, but I'm not sure if there's any particular reason for only using two in the first place?

Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/network/network_request_inspector_views.dart:423]: I figured that's what this is for - I've noticed that issue in the past. Adding a 3rd color sgtm

File: packages/devtools_app/lib/src/vm_service_wrapper.dart [hunk 11]
```diff
@@ -503,13 +503,22 @@ class VmServiceWrapper implements VmService {
...
+
   /// The `getHttpProfile` RPC is used to retrieve HTTP profiling information
   /// for requests made via `dart:io`'s `HttpClient`.
   ///
   /// The returned [HttpProfile] will only include requests issued after
   /// [httpTimelineLogging] has been enabled or after the last
   /// [clearHttpProfile] invocation.
-  Future<HttpProfile> getHttpProfile(String isolateId, {int updatedSince}) {
+  Future<HttpProfile> getHttpProfile(String isolateId,
+      {int updatedSince}) async {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/vm_service_wrapper.dart:520]: nit: trailing comma

File: packages/devtools_app/lib/src/vm_service_wrapper.dart [hunk 12]
```diff
@@ -520,7 +529,8 @@ class VmServiceWrapper implements VmService {
   }
 
   Future<HttpProfileRequest> getHttpProfileRequest(
-      String isolateId, int id) {
+      String isolateId, int id) async {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/vm_service_wrapper.dart:532]: nit: trailing comma

File: packages/devtools_app/macos/Flutter/GeneratedPluginRegistrant.swift [hunk 13]
```diff
@@ -2,6 +2,8 @@
 //  Generated file. Do not edit.
 //
 
+// clang-format off
```
Comment (kenzieschmoll) [inline packages/devtools_app/macos/Flutter/GeneratedPluginRegistrant.swift:5]: revert this change

**Review by kenzieschmoll — APPROVED** [review]

**Comment by kenzieschmoll** [comment]
lgtm_army

**Review by kenzieschmoll — APPROVED** [review]
